design/anova/anova.Rmd

```diff
@@ -0,0 +1,395 @@
+      m[i, ] <- 0
+    } else {
+      m[i, ] <- m[i, ] / m[i, i]
+    }
+  }
+  return(m)
+}
+
+normalize_mat <- function(m) {
+  m / rowSums(m^2)
+}
+
+mx <- model.matrix(fit)
+dl <- doolittle(t(mx) %*% mx)
+df_md(fit, dl[6:8, ])
+df_md(fit, normalize_mat(dl[6:8, ]))
+df_md(fit, dl[2, ]) # d.f. do not match, stat matches a lot ()
+```
+
+
+# Type II Estimable Functions
+
+According to the SAS/STAT® 13.1 User’s Guide, type II estimable can be achieved through
+
+For fixed effect F1:
+
+1. All columns of L associated with effects not containing F1 (except F1) are zero.
+2. The submatrix of L associated with effect F1 is $(X_1^\top M X_1)^{-}X_1^\top M X_1$
+3. Each of the remaining submatrices of L associated with an effect F2 that contains F1 is $(X_1^\top M X_1)^{-}X_1^\top M X_2$
+
+Note: $X^{-}$ is the generalized g2-inverse of $X$ if $X$ is a square matrix. And $(X^\top X)^{-}X^\top X$ is used
+in `PROC GLM` to represent `L` symbolically.
+And
+$X_0$ : columns of $X$ whose associated effects do not contain F1
+$X_1$ : columns of $X$ associated with F1
+$X_2$ : columns of $X$ associated with F2 that contains F1
+$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$.
+
+for the following mmrm model
+```{r}
+fit <- mmrm(FEV1 ~ ARMCD + AVISIT + ARMCD * AVISIT + ar1(AVISIT | USUBJID), data = fev_data)
+```
+
+the design matrix is
+
+```
+mx <- model.matrix(fit)
+```
+
+and the effect is
+
+| effect |  column in design matrix |
+|--------|--------------------------|
+|  mu    |        1                 |
+| ARMCD  |        2                 |
+| AVISIT |        3,4,5             |
+| ARMCD * AVISIT  |    6,7,8        |
+
+the effect `ARMCD * AVISIT` contains `ARMCD` and `AVISIT`.
+
+Let's calculate the `L` for `ARMCD` and `AVISIT`
+
+### ARMCD part
+```{r}
+x0 <- mx[, c(1, 3, 4, 5)]
+x1 <- mx[, c(2), drop = FALSE]
+x2 <- mx[, c(6, 7, 8)]
+
+l <- rep(0, 8) # initial values
+m <- diag(rep(1, nrow(x0))) - x0 %*% MASS::ginv(t(x0) %*% x0) %*% t(x0)
+l1_coef <- MASS::ginv(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x1
+l[c(2)] <- l1_coef
+
+l2_coef <- MASS::ginv(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x2
+l[c(6, 7, 8)] <- l2_coef
+
+df_md(fit, l) # very close !
+```
+
+### AVISIT part
+
+```{r}
+x0 <- mx[, c(1, 2)]
+x1 <- mx[, c(3, 4, 5), drop = FALSE]
+x2 <- mx[, c(6, 7, 8)]
+
+l <- matrix(0, nrow = 3, ncol = 8) # initial values
+m <- diag(rep(1, nrow(x0))) - x0 %*% MASS::ginv(t(x0) %*% x0) %*% t(x0)
+l1_coef <- MASS::ginv(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x1
+l[, c(3, 4, 5)] <- l1_coef
+
+l2_coef <- MASS::ginv(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x2
+l[, c(6, 7, 8)] <- l2_coef
+
+df_md(fit, l) # also close
+```
+
+### ARMCD * AVISIT
+
+```{r}
+x0 <- mx[, c(1, 2, 3, 4, 5)]
+x1 <- mx[, c(6, 7, 8), drop = FALSE]
+
+l <- matrix(0, nrow = 3, ncol = 8) # initial values
+m <- x0 %*% MASS::ginv(t(x0) %*% x0) %*% t(x0)
+l1_coef <- MASS::ginv(t(x1) %*% m %*% x1) %*% t(x1) %*% m %*% x1
+l[, c(6, 7, 8)] <- l1_coef
+
+df_md(fit, l) # close to type 3?
```

Comment:
depending on the reference level, it can be very close. SAS changes its reference level automatically but if the contrast is the same the result also is very close